src/throttler/throttler.ts

```diff
@@ -145,53 +155,18 @@ export abstract class Throttler<T, R> {
     if (!taskData) {
       throw new Error(`taskData.get(${cursorIndex}) does not exist`);
     }
-    const task = taskData.task;
-    const tname = this.taskName(cursorIndex);
-    const t0 = Date.now();
+    const promises = [this.executeTask(taskData, cursorIndex)];
+    if (taskData.timeoutMillis) {
+      promises.push(this.initializeTimeout(taskData, cursorIndex));
+    }
 
+    let result;
     try {
-      const result = await this.handler(task);
-      const dt = Date.now() - t0;
-      if (dt < this.min) {
-        this.min = dt;
-      }
-      if (dt > this.max) {
-        this.max = dt;
-      }
-      this.avg = (this.avg * this.complete + dt) / (this.complete + 1);
-
-      this.success++;
-      this.complete++;
-      this.active--;
-      if (taskData.wait) {
-        taskData.wait.resolve(result);
-      }
-      this.taskDataMap.delete(cursorIndex);
-      this.process();
+      result = await Promise.race(promises);
     } catch (err) {
-      if (this.retries > 0) {
-        if (taskData.retryCount < this.retries) {
-          taskData.retryCount++;
-          this.retried++;
-          await backoff(taskData.retryCount, this.backoff);
-          logger.debug(`[${this.name}] Retrying task`, tname);
-          return this.handle(cursorIndex);
-        }
-      }
-
-      this.errored++;
-      this.complete++;
-      this.active--;
-      if (taskData.retryCount > 0) {
-        logger.debug(`[${this.name}] Retries exhausted for task ${tname}:`, err);
-      } else {
-        logger.debug(`[${this.name}] Error on task ${tname}:`, err);
-      }
-      if (taskData.wait) {
-        taskData.wait.reject(err);
-      }
-      this.finish(err);
+      this.onTaskFailed(err, taskData, cursorIndex);
     }
+    this.onTaskFulfilled(result, taskData, cursorIndex);
```

Comment:
This should be part of the `try {...}`, so either `onTaskFailed` or `onTaskFulfilled` is executed.

Reply:
I don't think we want `this.onTaskFulfilled()` inside the `try` block, because we don't want thrown errors there to use the retry logic in the `catch` block (`this.onTaskFulfilled()` throwing an error would be an implementation error in code that calls the throttler).

But we should return from the function immediately after `this.onTaskFailed()`.

Reply:
Sg to me.

Reply:
Add `return;` after `this.onTaskFailed()`